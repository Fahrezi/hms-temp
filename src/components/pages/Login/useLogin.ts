import { zodResolver } from '@hookform/resolvers/zod';
import { AxiosResponse } from 'axios';
import { useForm } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import { z } from 'zod';

import { useAuth } from '@/hooks/useAuth';

import { loginRequest } from '@/services/auth.service';

import { LoginResponse, User } from '@/types/auth';

const loginSchema = z.object({
  email: z.string().email('Invalid email address'),
  password: z.string().min(6, 'Password must be at least 6 characters'),
});

type LoginFormInputs = z.infer<typeof loginSchema>;

export const useLogin = () => {
  const navigate = useNavigate();
  const loginMutation = loginRequest();
  const { login } = useAuth();

  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<LoginFormInputs>({
    resolver: zodResolver(loginSchema),
    mode: 'onBlur',
  });

  const onSubmit = async (data: LoginFormInputs) => {
    try {
      const response: AxiosResponse<LoginResponse> = await loginMutation.mutateAsync(data);

      if (response?.data) {
        const user: User = { ...response.data, roles: ['admin'] };

        login(user, 'xxxxxxxxxxxxx');
        void navigate('/dashboard');
      }
    } catch (error) {
      console.error('Login error:', error);
    }
  };

  return {
    register,
    handleSubmit,
    onSubmit,
    errors,
    isSubmitting: isSubmitting || loginMutation.isPending,
  };
};
