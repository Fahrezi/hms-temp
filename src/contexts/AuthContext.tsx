import { createContext, type ReactNode, useCallback, useEffect, useReducer } from 'react';

import { decryptData, encryptData } from '@/utils/secureAuth';
import { getLocalStorage, removeLocalStorage, setLocalStorage } from '@/utils/storage';
import { AuthAction, AuthContextType, AuthState, User } from '@/types/auth';

const initialState: AuthState = {
  user: null,
  token: null,
  isAuthenticated: false,
  isLoading: true,
};

const authReducer = (state: AuthState, action: AuthAction): AuthState => {
  switch (action.type) {
    case 'LOGIN':
      return { ...state, ...action.payload, isAuthenticated: true, isLoading: false };
    case 'LOGOUT':
      return { ...initialState, isLoading: false };
    case 'RESTORE':
      return { ...state, ...action.payload, isLoading: false };
    default:
      return state;
  }
};

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(authReducer, initialState);

  useEffect(() => {
    const storedData = getLocalStorage<string>('auth');
    if (storedData) {
      const parsedData = decryptData(storedData);
      if (parsedData) {
        dispatch({ type: 'RESTORE', payload: parsedData });
      }
    } else {
      dispatch({ type: 'RESTORE' });
    }
  }, []);

  useEffect(() => {
    if (!state.isLoading) {
      setLocalStorage('auth', encryptData(state));
    }
  }, [state]);

  const login = useCallback((user: User, token: string) => {
    dispatch({ type: 'LOGIN', payload: { user, token, isAuthenticated: true } });
  }, []);

  const logout = useCallback(() => {
    removeLocalStorage('auth');
    dispatch({ type: 'LOGOUT' });
  }, []);

  return <AuthContext.Provider value={{ ...state, login, logout }}>{children}</AuthContext.Provider>;
};

export default AuthContext;
