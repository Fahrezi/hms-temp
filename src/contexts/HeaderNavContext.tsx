import { BreadcrumbType, HeaderNavAction, HeaderNavContextType, HeaderNavState } from '@/types/headerNav';
import { createContext, type ReactNode, useCallback, useReducer } from 'react';

const initialState: HeaderNavState = {
  breadcrumb: [],
  title: '',
  isLoading: true,
};

const authReducer = (state: HeaderNavState, action: HeaderNavAction): HeaderNavState => {
  switch (action.type) {
    case 'CHANGE_TITLE':
      return { ...state, ...action.payload, title: action.payload?.title || '', isLoading: false };
    case 'CHANGE_BREADCRUMB':
      return { ...state, ...action.payload, breadcrumb: action.payload?.breadcrumb || [], isLoading: false };
    default:
      return state;
  }
};

const HeaderNavContext = createContext<HeaderNavContextType | undefined>(undefined);

export const HeaderNavProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(authReducer, initialState);

  const changeTitle = useCallback((title: string) => {
    dispatch({ type: 'CHANGE_TITLE', payload: { title } });
  }, []);

  const changeBreadcrumb = useCallback((breadcrumb: BreadcrumbType[]) => {
    dispatch({ type: 'CHANGE_BREADCRUMB', payload: { breadcrumb } });
  }, []);

  return <HeaderNavContext.Provider value={{ ...state, changeTitle, changeBreadcrumb }}>{children}</HeaderNavContext.Provider>;
};

export default HeaderNavContext;
