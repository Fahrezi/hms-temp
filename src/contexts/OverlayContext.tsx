import { OverlayAction, OverlayContextType, OverlayState } from '@/types/overlay';
import { createContext, type ReactNode, useCallback, useEffect, useReducer } from 'react';
import toast, { Toaster } from 'react-hot-toast';

const initialState: OverlayState = {
  isActive: false,
  notifText: '',
  notifType: null,
};

const authReducer = (state: OverlayState, action: OverlayAction): OverlayState => {
  switch (action.type) {
    case 'ACTIVATE_OVERLAY':
      return { ...state, ...action.payload, isActive: action.payload?.isActive || false };
    case 'ACTIVATE_NOTIF':
      return { ...state, ...action.payload, notifText: action.payload?.notifText || '', notifType: action.payload?.notifType || null, };
    default:
      return state;
  }
};

const OverlayContext = createContext<OverlayContextType | undefined>(undefined);

export const OverlayProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(authReducer, initialState);

  const activateOverlay = useCallback((status: boolean) => {
    dispatch({ type: 'ACTIVATE_OVERLAY', payload: { isActive: status } });
  }, []);

  const activateNotif = useCallback(({ notifText, notifType }: { notifText: string; notifType: 'success' | 'error' }) => {
    dispatch({ type: 'ACTIVATE_NOTIF', payload: { notifText, notifType } });
  }, []);


  useEffect(() => {
    if (!state.isActive) return;
    const { body, documentElement: html } = document;
    const prevOverflow = body.style.overflow;
    const prevPadRight = body.style.paddingRight;
    const scrollbar = window.innerWidth - html.clientWidth;

    body.style.overflow = "hidden";
    if (scrollbar > 0) body.style.paddingRight = `${scrollbar}px`;

    return () => {
      body.style.overflow = prevOverflow;
      body.style.paddingRight = prevPadRight;
    };
  }, [state.isActive]);

  useEffect(() => {
    if (state.notifText && state.notifType) {
      if (state.notifType === 'success') {
        toast.success(state.notifText);
      } else {
        toast.error(state.notifText);
      }
    }
  }, [state.notifText, state.notifType]);

  return (
    <OverlayContext.Provider value={{ ...state, activateOverlay, activateNotif }}>
      {children}
      {state.isActive && <div className="bg-black/40 w-screen h-screen fixed inset-0 top-0 left-0 z-40"></div>}
      <Toaster position="top-right" />
    </OverlayContext.Provider>
  );
};

export default OverlayContext;
