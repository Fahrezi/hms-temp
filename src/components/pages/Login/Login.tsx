import { useLogin } from './useLogin';

import styles from './Login.module.scss';

const Login = () => {
  const { register, handleSubmit, onSubmit, errors, isSubmitting } = useLogin();

  return (
    <div className={styles.login}>
      <div className={styles.login__card}>
        <h1 className={styles['login__card--title']}>Login</h1>
        <form onSubmit={handleSubmit(onSubmit)}>
          <div className={styles.login__field}>
            <label htmlFor='email' className={styles['login__field--label']}>
              Email
            </label>
            <input
              id='email'
              type='email'
              {...register('email')}
              className={styles['login__field--input']}
              disabled={isSubmitting}
            />
            {errors.email && <p className={styles['login__error']}>{errors.email.message}</p>}
          </div>

          <div className={styles.login__field}>
            <label htmlFor='password' className={styles['login__field--label']}>
              Password
            </label>
            <input
              id='password'
              type='password'
              {...register('password')}
              className={styles['login__field--input']}
              disabled={isSubmitting}
            />
            {errors.password && <p className={styles['login__error']}>{errors.password.message}</p>}
          </div>

          <button type='submit' className={styles.login__button} disabled={isSubmitting}>
            {isSubmitting ? 'Logging in...' : 'Login'}
          </button>
        </form>
        <p className={styles.login__register}>
          Don&apos;t have an account? <a href='/register'>Register now!</a>
        </p>
      </div>
    </div>
  );
};

export default Login;
