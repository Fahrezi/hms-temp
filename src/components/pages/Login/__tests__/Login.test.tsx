import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { vi } from 'vitest';

import Login from '@/components/pages/Login';

import { AuthProvider } from '@/contexts/AuthContext';

vi.mock('@/services/auth.service', () => ({
  loginRequest: () => ({
    mutateAsync: vi.fn().mockResolvedValue({
      data: {
        user: { id: 1, email: 'test@example.com', accessToken: 'fake-token' },
      },
    }),
  }),
}));

describe('Login Component', () => {
  it('renders the login component', () => {
    const { container } = render(
      <MemoryRouter>
        <AuthProvider>
          <Login />
        </AuthProvider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('handles login submission', async () => {
    render(
      <MemoryRouter>
        <AuthProvider>
          <Login />
        </AuthProvider>
      </MemoryRouter>,
    );

    const emailInput = screen.getByLabelText(/Email/i);
    const passwordInput = screen.getByLabelText(/Password/i);
    const loginButton = screen.getByRole('button', { name: /Login/i });

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });
    fireEvent.click(loginButton);

    await waitFor(() => {
      expect(screen.getByText(/Logging in.../i)).toBeInTheDocument();
    });
  });
});
