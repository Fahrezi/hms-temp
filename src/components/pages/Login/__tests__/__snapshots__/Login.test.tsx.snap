// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Component > renders the login component 1`] = `
<div>
  <div
    class="_login_499337"
  >
    <div
      class="_login__card_499337"
    >
      <h1
        class="_login__card--title_499337"
      >
        Login
      </h1>
      <form>
        <div
          class="_login__field_499337"
        >
          <label
            class="_login__field--label_499337"
            for="email"
          >
            Email
          </label>
          <input
            class="_login__field--input_499337"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div
          class="_login__field_499337"
        >
          <label
            class="_login__field--label_499337"
            for="password"
          >
            Password
          </label>
          <input
            class="_login__field--input_499337"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <button
          class="_login__button_499337"
          type="submit"
        >
          Login
        </button>
      </form>
      <p
        class="_login__register_499337"
      >
        Don't have an account? 
        <a
          href="/register"
        >
          Register now!
        </a>
      </p>
    </div>
  </div>
</div>
`;
